File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.cpp
#include <string>
#include <vector>
#include <map>
#include <set>
#include <iostream>
using namespace std;

bool solution(vector<string> phone_book) {
    bool answer = true;
    int cnt = 0;
    set<string> st;
    
    for(auto s : phone_book)st.insert(s);
    
    for(auto s : phone_book) {
        for(int i=0;i<s.size();++i) {
            if(st.find(s.substr(0,i+1)) != st.end()) cnt++;
        }
    }
        
    return (cnt == phone_book.size()) ? true : false;
}

File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    answer = True
    dic = {}
    for phone_num in phone_book:
        dic[phone_num] = dic.get(phone_num, phone_num)
    
    for phone_num in phone_book:
        for i in range(len(phone_num)):
            prefix = phone_num[:i]
            if dic.get(prefix):
                 return False
            
    return True